src/java/org/apache/cassandra/index/sai/utils/RangeIntersectionIterator.java

```diff
@@ -0,0 +1,229 @@
+        {
+            // Try advance all iterators to the highest key seen so far.
+            // Once this inner loop finishes normally, all iterators are guaranteed to be at the same value.
+            for (int index = 0; index < ranges.size(); index++)
+            {
+                if (index != alreadyAvanced)
+                {
+                    RangeIterator range = ranges.get(index);
+                    PrimaryKey nextKey = nextOrNull(range, highestKey);
+                    if (nextKey == null || nextKey.compareTo(highestKey) > 0)
+                    {
+                        // We jumped over the highest key seen so far, so make it the new highest key.
+                        highestKey = nextKey;
+                        // Remember this iterator to avoid advancing it again, because it is already at the highest key
+                        alreadyAvanced = index;
+                        // This iterator jumped over, so the other iterators are lagging behind now,
+                        // including the ones already advanced in the earlier cycles of the inner loop.
+                        // Therefore, restart the inner loop in order to advance
+                        // the other iterators except this one to match the new highest key.
+                        continue outer;
+                    }
+                    assert nextKey.compareTo(highestKey) == 0:
+                    String.format("skipTo skipped to an item smaller than the target; " +
+                                  "iterator: %s, target key: %s, returned key: %s", range, highestKey, nextKey);
+                }
+            }
+            // If we reached here, next() has been called at least once on each range iterator and
+            // the last call to next() on each iterator returned a value equal to the highestKey.
+            return highestKey;
+        }
+        return endOfData();
+    }
+
+    protected void performSkipTo(PrimaryKey nextToken)
+    {
+        for (RangeIterator range : ranges)
+            if (range.hasNext())
+                range.skipTo(nextToken);
+    }
+
+    /**
+     * Fetches the next available item from the iterator, such that the item is not lower than the given key.
+     * If no such items are available, returns null.
+     */
+    private PrimaryKey nextOrNull(RangeIterator iterator, PrimaryKey minKey)
+    {
+        iterator.skipTo(minKey);
+        return iterator.hasNext() ? iterator.next() : null;
+    }
+
+    public void close() throws IOException
+    {
+        ranges.forEach(FileUtils::closeQuietly);
+    }
+
+    public static Builder builder()
+    {
+        return new Builder();
+    }
+
+    public static Builder selectiveBuilder()
+    {
+        return selectiveBuilder(INTERSECTION_CLAUSE_LIMIT);
+    }
+
+    public static Builder selectiveBuilder(int limit)
+    {
+        return new Builder(limit);
+    }
+
+    public static class Builder extends RangeIterator.Builder
```

Comment:
nit: I'd probably reduce the visibility of `Builder` and its constructors at least to package-private, given we should always be using the static factories.

Reply:
I guess that applies to the other iterator implementations as well.